File: CompetitiveProgramming/DMOJ/TLE/TLE '15 P1 - Power Rankings/2021-07-25/PowerRanking.java
/*
#####################################################################################
############################ TLE '15 P1 - Power Rankings ############################
Output the rankings of names based on number of minutes viewing PowerPoints.
The ranking is ordered by 3 + students with a greater total time viewing PowerPoints.

Sample Input
5 3
EnochPoon
LiuJason
Nathanl3
AlexW
TT1103
3 4 2 1 0
1 3 4 2 0
5 1 7 4 0

Sample Input Explanation
L1: n p, number of names, number of PowerPoints.
L2 to Ln+1: names.
Ln+1+1 to Ln+1+p: t[0] t[1] ... t[n], time the ith student spends on a PowerPoint.

Sample Output
3. Nathanl3
4. EnochPoon
5. LiuJason
6. AlexW
7. TT1103

########################################################################
############################ SOLUTION LOGIC ############################
Pair time and name of each person.
Sort by time in reverse order.
Output starting from 3.
*/

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;

/**
 * Program to find the "power ranking".
 * @author Vincent Zhang
 * @since 25 July 2021
 */
public class PowerRanking
{
    /**
     * Class to store the data including the sum and name of each person.
     * @author Vincent Zhang
     * @since 25 July 2021
     */
    class Data implements Comparable<Data>
    {
        int sum;
        String name;

        /**
         * Initialize values upon instantiation.
         * @param sum Sum of the time spent reading PowerPoints
         * @param name Name of the person
         */
        public Data(int sum, String name)
        {
            this.sum = sum;
            this.name = name;
        }

        /**
         * Comparable method used by sort, to sort data in reverse order.
         * @param d Data to compare against
         * @return If d.sum < this.sum, return a value < 0
         */
        @Override
        public int compareTo(Data d)
        {
            return Integer.compare(d.sum, this.sum);
        }
    }

    // Global declarations.
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st;

    /**
     * Main method to solve the problem.
     * @return void
     */
    void run()
    {
        int n = nextInt();
        int p = nextInt();

        // Get the data including the sum of time spent reading PowerPoints and names of each person.
        List<Data> info = new ArrayList<>();
        for(int i=0; i<n; i++)
            info.add(new Data(0, nextLine()));
        for(int i=0; i<p; i++)
        {
            for(int j=0; j<n; j++)
            {
                int time = nextInt();
                info.set(j, new Data(info.get(j).sum + time, info.get(j).name));
            }
        }

        // Sort and output data in order of the "power ranking".
        Collections.sort(info);
        for(int i=0; i<n; i++)
            System.out.println((i+3) + ". " + info.get(i).name);
    }

    /**
     * Instantiates PowerRanking to reference from non-static context.
     * @param args Command-line arguments
     * @return void
     */
    public static void main(String[] args)
    {
        PowerRanking pr = new PowerRanking();
        pr.run();
    }

    /**
     * Find the next token in the input string.
     * @return Next token in input string
     */
    String nextToken()
    {
        while(st == null || !st.hasMoreTokens())
            try {
                st = new StringTokenizer(br.readLine());
            } catch(IOException e) {e.printStackTrace();}
        return st.nextToken();
    }

    /**
     * Find the next token in integer value.
     * @return Integer value of the next token of the input string
     */
    int nextInt()
    {
        int n = -1;
        try {
            n = Integer.parseInt(nextToken());
        } catch(NumberFormatException e) {e.printStackTrace();}
        return n;
    }

    /**
     * Read entire line.
     * @return Entire string line of input
     */
    String nextLine()
    {
        String s = "";
        try {
            s = br.readLine();
        } catch(IOException e) {e.printStackTrace();}
        return s;
    }

}

File: CompetitiveProgramming/DMOJ/TLE/TLE '15 P1 - Power Rankings/2021-07-25/PowerRanking.cpp
/*
#####################################################################################
############################ TLE '15 P1 - Power Rankings ############################
Output the rankings of names based on number of minutes viewing PowerPoints.
The ranking is ordered by 3 + students with a greater total time viewing PowerPoints.

Sample Input
5 3
EnochPoon
LiuJason
Nathanl3
AlexW
TT1103
3 4 2 1 0
1 3 4 2 0
5 1 7 4 0

Sample Input Explanation
L1: n p, number of names, number of PowerPoints.
L2 to Ln+1: names.
Ln+1+1 to Ln+1+p: t[0] t[1] ... t[n], time the ith student spends on a PowerPoint.

Sample Output
3. Nathanl3
4. EnochPoon
5. LiuJason
6. AlexW
7. TT1103

########################################################################
############################ SOLUTION LOGIC ############################
Pair time and name of each person.
Sort by time in reverse order.
Output starting from 3.
*/

/**
 * Program to find the "power rankings".
 * @author Vincent Zhang
 * @since 25 July 2021
 */
#include <iostream>
#include <vector>
#include <algorithm>

/**
 * Public class to store time sum and name.
 * @author Vincent Zhang
 * @since 25 July 2021
 */
struct Data
{
    int sum;
    std::string name;
}
info[101]; // Data info[101].

/**
 * Compare method to sort Data in reverse order.
 * @param a First person sum
 * @param b Second person sum
 * @return Whether first person sum > second person sum
 */
bool compare(Data a, Data b)
{
    return a.sum > b.sum;
}

/**
 * Main function to solve the problem.
 * @param argc Number of command-line arguments + 1 for program name
 * @param argv Program name and command-line arguments
 */
int main(int argc, char *argv[])
{
    // Disable synchronization between C and C++ standard streams, allowing C++ streams independent buffers.
    std::ios_base::sync_with_stdio(false);
    // Untie cin stream from cout stream to disable automatic flushing after each I/O operation on the other stream.
    std::cin.tie(nullptr);

    // Store names.
    int n, p; std::cin >> n >> p;
    for(int i=0; i<n; i++)
    {
        std::string name; std::cin >> name;
        info[i].sum = 0;
        info[i].name = name;
    }

    // Store sum of times.
    for(int i=0; i<p; i++)
    {
        for(int j=0; j<n; j++)
        {
            int t; std::cin >> t;
            info[j].sum += t;
        }
    }

    // Output result.
    std::sort(info, info+n, compare);
    for(int i=0; i<n; i++)
        std::cout << i+3 << ". " << info[i].name << '\n';

    return 0;
}